source/agora/common/Config.d

```diff
@@ -0,0 +1,310 @@
+/*******************************************************************************
+
+    Define the configuration objects that are used through the application
+
+    See `doc/config.example.yaml` for some documentation.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Config;
+
+import agora.common.crypto.Key;
+import agora.common.Set;
+
+import dyaml.node;
+
+import vibe.core.log;
+
+import std.algorithm;
+import std.conv;
+import std.exception;
+import std.format;
+import std.range;
+import std.traits;
+
+/// Verifies the config is in a valid state.
+/// This function should be called both after parsing a config file,
+/// or after manually constructing a config file.
+public void verifyConfigFile ( Config config )
+{
+    // ensure a quorum doesn't contain a public key of this node (self-reference)
+    Set!PublicKey all_validators;
+    getAllValidators(config.quorums, all_validators);
+    enforce(config.node.key_pair.address !in all_validators,
+        format("Cannot have our own key as a validator node: %s", config.node.key_pair.address));
+}
+
+// extract the set of all validators as configured in the quorum config
+public void getAllValidators ( in QuorumConfig[] quorums,
+    ref Set!PublicKey result )
+{
+    foreach (quorum; quorums)
+    {
+        result.fill(quorum.nodes);
+        getAllValidators(quorum.quorums, result);
+    }
+}
+
+/// Command-line arguments
+public struct CommandLine
+{
+    /// Path to the config file
+    string config_path = "config.yaml";
+
+    /// check state of config file and exit early
+    bool config_check;
+}
+
+/// Main config
+public struct Config
+{
+    static assert(!hasUnsharedAliasing!(typeof(this)));
```

Comment:
Add a problem explaining the comment why

Reply:
_because immutable sucks_ :PP